Find a longer PV within a certain range if the best PV is too short.

lakinwecker opened the pull request:
If the best chosen PV is shorter than a given number of plies, look for
a better PV to use which is within a certain score and depth range.
Ensure that the longer PV uses all of the same moves from the chosen PV.

Send it to the GUI after the chosen PV is sent.

This likely needs a bit of tuning and careful consideration, but it
should find a reasonable PV with perhaps a bit of loss of strength.

It probably needs to be heavily tested on the games that show the small
PVs to ensure that the chosen PV is worth showing.

lakinwecker committed:
Find a longer PV within a certain range if the best PV is too short.

If the best chosen PV is shorter than a given number of plies, look for
a better PV to use which is within a certain score and depth range.
Ensure that the longer PV uses all of the same moves from the chosen PV.

Send it to the GUI after the chosen PV is sent.

This likely needs a bit of tunining and careful consideration, but it
should find a reasonable PV with perhaps a bit of loss of strength.

It probably needs to be heavily tested on the games that show the small
PVs to ensure that the chosen PV is worth showing.

ornicar commented:
/cc @ddugovic can we get a code review from you pretty please :) :) :)

ddugovic commented:
Nice algorithm although I'm concerned about a couple details:

```
// Look for the best thread that meets the minimum move criteria.
```

I don't know whether thread-blocking or race conditions are reasonable concerns here (as other threads could change their PVs in the middle of executing this code).

Also, I think there are three types of evaluations: upper-bound, lower-bound, and score.  So it might not be possible to compare solely on `eval`?

Maybe a safer approach would be to create a new command (maybe `pv`) executed after Stockfish outputs `bestmove` and stops searching, which could extract from the transposition table a PV using this algorithm (or one like it).

lakinwecker commented:
Thanks for the feedback. Good stuff.

> I don't know whether thread-blocking or race conditions are reasonable concerns here (as other threads could change their PVs in the middle of executing this code).

Perhaps my assumptions are wrong, but I assumed that the threads were done searching based on this: https://github.com/niklasf/Stockfish/pull/1/files#diff-71c8faa29f847041392f6e81e16bd9e6L375
Which happens directly before this and should mean that the mainThread (where this algorithm is running) has stopped searching, and has stopped all threads from searching and is now ready to find and print out the best line.  My interpretation of this is that there should be no race conditions at this point.

This is further supported by this code, which is what I was patterning mine after: https://github.com/niklasf/Stockfish/pull/1/files#diff-71c8faa29f847041392f6e81e16bd9e6L387

In this case, they are using the score and depth directly to find the "best" line from all threads.  I am doing the same thing, except that I'm allowing for a slightly worse line (both in terms of depth/score) so long as it pushes the length of the principle variation past the threshold.  My interpretation of that code is that comparing scores directly should be fine. Stockfish is already doing this. I'm just tweaking the comparison to find a longer line. 

All of this is motivated by something @ornicar said today regarding rolling back stockfish to a version that didn't use short variations even if the elo was dropped 100 points.  I thought I could effectively do the same thing (provide a slightly less strong PV), but perhaps still benefit from most of the increased strength from the newer stockfish.

I also agree that this is not a long term solution. At some point in the past, there was a method called `extract_pv_from_tt` in that file, that seems to do what we want. It's been replaced with `extract_ponder_from_tt` - but I am going to see if I can resurrect it and use it to find the actual PV that was used to choose the move/best eval.  However, it might be worth using this temporarily to work around the short variations until we have that in place.   In that case, I'm happy to make this a separate command or have it just always ensure that the PV is longer for the chosen move.

ddugovic commented:
Indeed, I like this concept (for a shorter-term patch) & I think your analysis is correct.

lakinwecker commented:
Great! 

It was pointed out in irc that I may have made this PR against the wrong repo/branch.  I'm happy to fix that and make it against the right branch/repo if you'd like.  Just point me in the right direction.

ianfab commented on src/search.cpp:
```diff
@@ -387,7 +391,33 @@ void MainThread::search() {
+              }
+              if (
+                     ((th->rootMoves[0].pv.size() >= minPlies)
+                  ||  (abs(th->rootMoves[0].score - longestPVThread->rootMoves[0].score) > maxScoreDiff))
```
I would expect a less-than sign here. Is it a typo or intended?

lakinwecker replied:
Thanks for this. I believe you are correct, in this case it should have been a less-than - but I also believe that the entire if statement is not doing what I had intended so I'm going to see if I can write it in a more clear manner and ensure it's doing what I wanted it to.

ianfab commented on src/search.cpp:
```diff
@@ -387,7 +391,33 @@ void MainThread::search() {
+              }
+              bool possiblePVReplacement = true;
+              for (unsigned int i = 0; i < longestPVThread->rootMoves[0].pv.size(); ++i) {
+                  if (th->rootMoves[0].pv[i] != longestPVThread->rootMoves[0].pv[i]) {
```
Would it not be sufficient to compare the PV of the thread with `bestThread`? By comparing with `longestPVThread` the result depends on the order in which the threads are looped over in case there is more than one thread that meets all criteria, but where the PVs are different (while the first moves are agreeing with the PV of `bestThread`).

lakinwecker replied:
Yes, you are correct. Here we want to compare to bestThread.

lakinwecker committed:
Clarify search criteria, prefer higher eval once our PV is long enough

In addition to making the algorithm more readable (IMO), this fixes a
couple of issues raised during the review.

1. any PV that matches the bestThread PV is a candidate now. Previously
I was mistakenly requiring each subsequent thread to match the previous
longestPVThread, which was dependent on the order the threads showed
up. This is much better.

2. So long as our PV is smaller than the minPlies, we will take any
thread that provides a longer candidate PV and a score/depth that is
"close enough" to the bestThread.  Once our PV has reached minPlies, we
will take any thread that improves either score or depth, so long as it
also is minPlies long.

I think this is a nice middle ground, although the values for minPlies,
maxScoreDiff and maxDepthDiff should be tuned. I just chose 10
centipawns for score and 2 for depth based on some assumptiong I
personally made, but I wouldn't argue for them at all.  minPlies was
set to 6 based on feedback from Thibault, as ultimately we are looking
to satifisy Users perception in the UI, and so I wanted to ensure I was
fixing that.

Also note that this is not guaranteed to provide a longer PV. If all
threads have short PVs (not sure if that will ever happen) then we may
still get a shorter PV.  I hope to be able to work on the Transposition
Table PV extraction over the holidays.

lakinwecker commented:
Thanks for the review @ianfab. Based on your feedback I have just pushed a new change - which I believe addresses both of your points and makes the algorithm better.  Here is my commit message, which is an attempt to describe what I did:

In addition to making the algorithm more readable (IMO), this fixes a
couple of issues raised during the review.

1. any PV that matches the bestThread PV is a candidate now. Previously
I was mistakenly requiring each subsequent thread to match the previous
longestPVThread, which was dependent on the order the threads showed
up. This is much better.

2. So long as our PV is smaller than the minPlies, we will take any
thread that provides a longer candidate PV and a score/depth that is
"close enough" to the bestThread.  Once our PV has reached minPlies, we
will take any thread that improves either score or depth, so long as it
also is minPlies long.

I think this is a nice middle ground, although the values for minPlies,
maxScoreDiff and maxDepthDiff should be tuned. I just chose 10
centipawns for score and 2 for depth based on some assumptions I
personally made, but I wouldn't argue for them at all.  minPlies was
set to 6 based on feedback from Thibault, as ultimately we are looking
to satifisy Users perception in the UI, and so I wanted to ensure I was
fixing that.

Also note that this is not guaranteed to provide a longer PV. If all
threads have short PVs (not sure if that will ever happen) then we may
still get a shorter PV.  I hope to be able to work on the Transposition
Table PV extraction over the holidays.

lakinwecker commented:
Also, I need to do more testing on further positions. I'm going to speak with @ornicar about that now. 

This is the test position I have been using so far. https://en.lichess.org/jNsrRl4a#16

With these stockfish commands:
```uci
setoption name hash value 512
setoption name threads value 4
isready
position fen rnbqkb1r/1ppp1ppp/8/p1P1P3/P2p4/2N5/1P1P1PPP/R1BQK1NR w KQkq - 0 9
isready
go nodes 3500000
```

This is the example output which shows the short variation followed by my longer variation:

```
info depth 1 seldepth 1 multipv 1 score cp -48 nodes 5503 nps 275150 tbhits 0 time 20 pv c3e4 b8c6 d2d3
info depth 2 seldepth 2 multipv 1 score cp -47 nodes 5538 nps 276900 tbhits 0 time 20 pv c3e4 d4d3
info depth 3 seldepth 3 multipv 1 score cp -97 nodes 5593 nps 279650 tbhits 0 time 20 pv c3e4 d4d3 f2f3 b8a6
info depth 4 seldepth 4 multipv 1 score cp -97 nodes 5657 nps 282850 tbhits 0 time 20 pv c3e4 d4d3 f2f3 b8a6
info depth 5 seldepth 5 multipv 1 score cp -114 nodes 5992 nps 299600 tbhits 0 time 20 pv c3e4 d4d3 g1f3 b8a6 b2b3 a6c5 e4c5
info depth 6 seldepth 6 multipv 1 score cp -74 nodes 6367 nps 303190 tbhits 0 time 21 pv c3e4 d4d3 d1h5 b8c6 g1f3 g7g6
info depth 7 seldepth 7 multipv 1 score cp -45 nodes 7530 nps 301200 tbhits 0 time 25 pv c3e4 d8e7 d2d3 b8a6 c1g5 e7e5 g1f3
info depth 8 seldepth 8 multipv 1 score cp -53 nodes 16131 nps 384071 tbhits 0 time 42 pv c3e4 d7d5 e5d6 c7d6 d2d3 h7h6 c5d6 f8d6
info depth 9 seldepth 13 multipv 1 score cp -27 nodes 20100 nps 402000 tbhits 0 time 50 pv c3e4 b8c6 g1f3 d8e7 d2d3 f7f6 e1g1 f6e5 c1g5
info depth 10 seldepth 13 multipv 1 score cp -32 nodes 44418 nps 499078 tbhits 0 time 89 pv c3e4 b8c6 g1f3 d7d6 e5d6 c7d6 c5d6 f8d6 e1g1 d6b4
info depth 11 seldepth 14 multipv 1 score cp -24 nodes 81636 nps 578978 tbhits 0 time 141 pv c3e4 b8a6 d2d3 f8c5 g1f3 d7d6 c1g5 c5b4 e4d2 d8d7 e1g1 d6e5 f3e5
info depth 12 seldepth 20 multipv 1 score cp -49 nodes 264079 nps 841015 tbhits 0 time 314 pv c3e4 b8a6 d2d3 f8c5 c1g5 c5b4 e1f1 b4e7 h2h4 e8g8 g1f3 h7h6 g5e7 d8e7
info depth 13 seldepth 20 multipv 1 score cp -59 nodes 314332 nps 890458 tbhits 0 time 353 pv c3e4 b8a6 d2d3 f8c5 g1f3 d7d6 c1g5 d8d7 e5d6 c7d6 d1e2 d7e6 e4c5 d6c5
info depth 14 seldepth 20 multipv 1 score cp -70 nodes 439836 nps 1008798 tbhits 0 time 436 pv c3e4 b8a6 d2d3 f8c5 c1g5 c5b4 e1f1 b4e7 h2h4 e8g8 g1f3 h7h6 g5e7 d8e7 d1b3 e7b4 b3b4 a6b4
info depth 15 seldepth 23 multipv 1 score cp -91 nodes 845103 nps 1332970 tbhits 0 time 634 pv c3e4 b8a6 d2d3 f8c5 g1f3 d7d6 c1g5 c5b4 e1f1 d8d7 e5d6 b4d6 e4d6 d7d6 d1e1 e8f8 e1e4 a6c5 e4d4 d6d4 f3d4
info depth 16 seldepth 28 multipv 1 score cp -93 nodes 1101479 nps 1480482 tbhits 0 time 744 pv c3e4 b8a6 d2d3 f8c5 g1f3 d7d6 c1g5 c5b4 e1f1 d8d7 e5d6 c7d6 h2h4 d6d5 e4g3 e8g8 f3d4 f8e8 g5d2 a6c5 d2b4 a5b4 g3f5 a8a4
info depth 17 seldepth 28 multipv 1 score cp -106 nodes 2241913 nps 1998139 hashfull 28 tbhits 0 time 1122 pv c3e4 b8a6 d2d3 f8c5 g1f3 e8g8 c1g5 c5b4 e1f1 d8e8 g5f4 a6c5 f3d4 d7d6 e4c5 d6e5 d4b5 e5f4 b5c7 e8e7 c7a8
info depth 18 seldepth 30 multipv 1 score cp -98 nodes 3514264 nps 2553970 hashfull 51 tbhits 0 time 1376 pv c3e4 b8a6
info depth 17 seldepth 28 multipv 1 score cp -97 nodes 3514264 nps 2553970 hashfull 51 tbhits 0 time 1376 pv c3e4 b8a6 d2d3 f8c5 g1f3 e8g8 e1g1 b7b6 c1g5 c5e7 g5e7 d8e7 f1e1 a6b4 f3d4 e7e5 d4f3 e5d5 a1c1 c8b7 c1c7
bestmove c3e4 ponder b8a6
```

ianfab commented:
Thanks @lakinwecker for the update and for adressing my points. However, in my opinion, it would be better to only consider threads where the score is within a certain range *in both directions*, because otherwise I think there could be incorrect PVs when failing high/low deeper in the search tree, and the score might be off by a huge margin.

lakinwecker commented:
Stockfish already goes through all of the threads and takes the thread where `th->rootMoves[0]` has the best score/depth as its `bestThread`.   https://github.com/niklasf/Stockfish/pull/1/files#diff-71c8faa29f847041392f6e81e16bd9e6R391 If these fail high/low pvs are showing up in one of the `th->rootMoves[0]`, then stockfish is already using them as its best thread - this does not change that. 

In my algorithm, we will only ever replace the `bestThread` if it's not long enough.  Because we already know that this thread has the best score of any of the threads, this patch will only weaken the score. Which we only do while lengthening the `pv` to meet the `minPlies` requirement.  After that, when taking threads with a better score, we can never reach the `bestThread` score again, because we know it's the max score within our search space and that it was not long enough to begin with. 

Maybe I am misunderstanding your point?

lakinwecker commented:
A better explanation is that this change only evaluates the same threads that Stockfish already evaluated to choose its `bestThread`. As such, this will only ever produce a PV that is within `bestThread->rootMoves[0].score` and `bestThread->rootMoves[0].score - maxScoreDiff`

ianfab commented:
It only updates `bestThread` if both `score` and `completedDepth` of the thread are higher than that of the previous `bestThread`, so a thread with lower (or equal) depth but higher score should be possible, right?

lakinwecker commented:
Good call, you are correct. There is the possibility for an equal or greater depth but higher score that's not considered the bestThread is possible. I can add the extra criteria to keep the score within that +- maxScoreDiff

lakinwecker committed:
limit the score diff both ways so that we aren't choosing a wildly different pv that may have failed high/low

lakinwecker committed:
limit the score diff both ways so that we aren't choosing a wildly different pv that may have failed high/low

lakinwecker committed:
use abs value to compare both ways

lakinwecker commented:
`maxScoreDiff` is now enforced in both directions.

lakinwecker committed:
Merge branch 'master' of github.com:ddugovic/Stockfish into ddugovic
